package/MDAnalysis/analysis/bat.py

```diff
@@ -0,0 +1,503 @@
+            (p0, p1, p2) = self._root.positions
+        else:
+            (p0, p1, p2) = make_whole(self._root, inplace=False)
+        v01 = p1 - p0
+        v21 = p1 - p2
+        # Internal coordinates
+        r01 = np.sqrt(np.sum(v01 *
+                             v01))  # Distance between first two root atoms
+        r12 = np.sqrt(np.sum(v21 *
+                             v21))  # Distance between second two root atoms
+        a012 = np.arccos(max(-1.,min(1.,np.sum(v01*v21)/\
+          np.sqrt(np.sum(v01*v01)*np.sum(v21*v21))))) # Angle between root atoms
+        # Exernal coordinates
+        e = v01 / r01
+        phi = np.arctan2(e[1], e[0])  # Polar angle
+        theta = np.arccos(e[2])  # Azimuthal angle
+        # Rotation to the z axis
+        cp = np.cos(phi)
+        sp = np.sin(phi)
+        ct = np.cos(theta)
+        st = np.sin(theta)
+        Rz = np.array([[cp * ct, ct * sp, -st], [-sp, cp, 0],
+                       [cp * st, sp * st, ct]])
+        pos2 = Rz.dot(p2 - p1)
+        # Angle about the rotation axis
+        omega = np.arctan2(pos2[1], pos2[0])
+        root_based = np.concatenate((p0, [phi, theta, omega, r01, r12, a012]))
+
+        # Calculate internal coordinates from the torsion list
+        bonds = calc_bonds(self._ag1.positions,
+                           self._ag2.positions,
+                           box=self._ag1.dimensions)
+        angles = calc_angles(self._ag1.positions,
+                             self._ag2.positions,
+                             self._ag3.positions,
+                             box=self._ag1.dimensions)
+        torsions = calc_dihedrals(self._ag1.positions,
+                                  self._ag2.positions,
+                                  self._ag3.positions,
+                                  self._ag4.positions,
+                                  box=self._ag1.dimensions)
+        # When appropriate, calculate improper torsions
+        shift = torsions[self._primary_torsion_indices]
+        shift[self._unique_primary_torsion_indices] = 0.
+        torsions -= shift
+        # Wrap torsions to between -np.pi and np.pi
+        torsions = ((torsions + np.pi) % (2 * np.pi)) - np.pi
+
+        self.bat.append(np.concatenate((root_based, bonds, angles, torsions)))
+
+    def load_bat(self, FN):
+        """Loads the bat trajectory from a netcdf file
+        """
+        nc_F = Dataset(FN,'r')
+        self.bat = list(np.array(nc_F.variables['bat']))
+        nc_F.close()
+
+    def save_bat(self, FN):
+        """Saves the bat trajectory to a netcdf file
+        """
+        bat = np.array(self.bat)
+
+        nc_F = Dataset(FN,'w')
+        frames = nc_F.createDimension("frames", bat.shape[0])
+        dims = nc_F.createDimension("dims", bat.shape[1])
+        bat_in_F = nc_F.createVariable("bat", "f4", ("frames","dims"))
+        bat_in_F[:] = bat
+        nc_F.close()
+
+    def Cartesian(self, bat):
```

Comment:
I find it odd that you pass `bat` in here. I think this should just be
```python
def Cartesian(self):
   bat = self.bat
   origin = bat[:3]
   ...
   return XYZ
```
and then use `self.bat` (the object knows about its data).

You are using `self._primary_torsion_indices` and `self._unique_primary_torsion_indices` so I don't see how you could make it a stand-alone function.

Reply:
This function is really for interfacing with other modules. For example, you may want to change one dihedral angle and get a new set of Cartesian coordinates.

Reply:
You mean you take `R.bat`, change something, and then want to get new Cartesian coordinates back without changing the original bat?

In this case I'd do the following
- make `bat=None` optional and default to `self.bat`
- add a sanity check that `bat` can be used (at least dimensions)
- write more docs, outlining what the method is good for

I assume you can't make it a function, independent of the `BAT` class?